### PR TITLE
Properties window

### PR DESCRIPTION
**Pull request by Tomer51m**
Added Properties window component with the first tow tabs.
At the moment there's are bug regarding the state management in the Desktop tab. 
Working on fixing it.

### CODE REVIEW

File: src/components/ContextMenu/index.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    if (displayFocus || parentRef)
+      setRefPos(elementRef.current.getBoundingClientRect());
+  }, [displayFocus, parentRef]);
+
+  const initializeMenuPosition = useCallback(
+    (mousePos, menuRef, parentRef, isToLeft) => {
+      const windowWidth = window.innerWidth;
+      const windowHeight = window.innerHeight;
+      const offSetX = 10;
+      const offSetY = 20;
```
Comment (ShaharTal) [inline src/components/ContextMenu/index.js:39]: Change offSet to offset, it is the same word :)

Comment (Tomer51m) [reply src/components/ContextMenu/index.js:39]: Correct. Fixed ✌️.

**Review by RonnyO — CHANGES_REQUESTED** [review]
Awesome! Left some comments :-)

File: src/WinXP/apps/index.js [hunk 2]
```diff
@@ -16,6 +17,7 @@ import notepadLarge from 'assets/windowsIcons/327(32x32).png';
 import winamp from 'assets/windowsIcons/winamp.png';
 import paintLarge from 'assets/windowsIcons/680(32x32).png';
 import paint from 'assets/windowsIcons/680(16x16).png';
+import empty from 'assets/empty.png';
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/index.js:20]: Done.

File: src/WinXP/apps/DisplayProperties/index.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import TabsMenu from '../TabsMenu';
+
+import ThemeTab from './ThemeTab';
+import AppearanceTab from './AppearanceTab';
+import DesktopTab from './DesktopTab';
+import ScreenSaverTab from './ScreenSaverTab';
+import SettingsTab from './SettingsTab';
+import { DESKTOP, CHANGE, DISPLAY_PROPERTIES } from './utils';
+
+import Button from '../../../components/Button';
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/index.js:15]: Done.

File: src/WinXP/apps/DisplayProperties/index.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+const tabs = [
+  { id: 1, title: 'Themes', content: ThemeTab },
+  { id: 2, title: 'Desktop', content: DesktopTab },
+  { id: 3, title: 'Screen Saver', content: ScreenSaverTab },
+  { id: 4, title: 'Appearance', content: AppearanceTab },
+  { id: 5, title: 'Settings', content: SettingsTab },
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/index.js:22]: Done.

File: src/WinXP/apps/DisplayProperties/index.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  useEffect(() => {
+    dispatch({
+      type: CHANGE,
+      payload:
+        JSON.stringify(state.displayProperties) !==
+        JSON.stringify(displayProperties),
+    });
+  }, [displayProperties, state.displayProperties]);
+
+  const handleCancel = () => onClose();
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/index.js:66]: Done.

File: src/WinXP/apps/DisplayProperties/index.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+  const handleApply = () =>
+    appContext.dispatch({
+      type: DISPLAY_PROPERTIES,
+      payload: { ...state.displayProperties },
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/index.js:71]: Done.

File: src/WinXP/apps/DisplayProperties/utils.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  { title: 'Santa Fe Stucco', background: santaFeStucco },
+  { title: 'Soap Bubbles', background: soapBubbles },
+  { title: 'Zapotec', background: zapotec },
+];
+
+const initializeBackgrounds = backgroundsObj => {
+  return backgroundsObj.map((item, index) => ({ ...item, id: index + 1 }));
+};
+
+export const backgrounds = initializeBackgrounds(backgroundsObj);
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/utils.js:72]: Done.

File: src/WinXP/apps/DisplayProperties/ThemeTab.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+const defaultDesktop = {
+  id: 5,
+  position: 'stretch',
+  image: bliss,
+  color: '#2f71cd',
+};
+
+function ThemeTab({ state, dispatch }) {
+  const [preview, setpreview] = useState(null);
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/ThemeTab.js:24]: Done.

File: src/WinXP/apps/DisplayProperties/ThemeTab.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  const handleSelectChange = e => {
+    const { value } = e.target;
+
+    switch (value) {
+      case classicXP:
+        setpreview(classicXP);
+        break;
+      case 'current':
+        setpreview(null);
+        const displayProps = getLocalStorage('display properties');
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/ThemeTab.js:35]: Done.

File: src/WinXP/apps/DisplayProperties/ThemeTab.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    }
+  };
+
+  return (
+    <ThemesPage>
+      <div className="description">
+        <p>
+          A theme is a background plus a set of sounds, icons and other elements
+        </p>
+        <p> to help you personalize your computer with one click.</p>
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/ThemeTab.js:57]: Done.

File: src/WinXP/apps/DisplayProperties/ThemeTab.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  .sample {
+    margin-top: 20px;
+
+    .preview {
+      position: relative;
+      margin-top: 5px;
+      height: 230px;
+      box-shadow: 0px 0px 3px 1px grey inset;
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/ThemeTab.js:150]: Done.

File: src/WinXP/apps/DisplayProperties/DesktopTab.js [hunk 12]
```diff
@@ -0,0 +1,257 @@
+import React, { useState, useEffect } from 'react';
+import styled from 'styled-components';
+
+import { backgrounds, DESKTOP } from './utils';
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/DesktopTab.js:4]: Done.

File: src/WinXP/apps/DisplayProperties/DesktopTab.js [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    color,
+  });
+
+  const refs = backgrounds.reduce((acc, item) => {
+    acc[item.id] = React.createRef();
+    return acc;
+  }, {});
+
+  useEffect(() => {
+    if (desktopState.id === 0) setDisablePosition(true);
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/DesktopTab.js:31]: Done.

File: src/WinXP/apps/DisplayProperties/DesktopTab.js [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    setInit(true);
+  }, [id, init, refs]);
+
+  // Unmount only.
+  useEffect(() => () => setInit(false), []);
+
+  const handleBackgroundClick = (e, id, background) => {
+    setDesktopState(prev => ({ ...prev, id }));
+
+    if (e.target.innerText === '(None)') {
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/DesktopTab.js:52]: It's not working as expected when I compare with id. I switched to compare string with the first item in the Backgrounds array.

File: src/WinXP/apps/DisplayProperties/DesktopTab.js [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+              >
+                <option value="center">Center</option>
+                <option value="tile">Tile</option>
+                <option value="stretch">Stretch</option>
+              </select>
+            </div>
+            <div>
+              <label htmlFor="color">Color:</label>
+              <input
+                id="color"
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/DesktopTab.js:119]: done.

File: src/WinXP/apps/DisplayProperties/DesktopTab.js [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+      padding: 0 2px;
+    }
+
+    & .active {
+      color: #fff;
+      background-color: #2f71cd;
+      border: 1px dotted grey;
+    }
+  }
+  .options {
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/DisplayProperties/DesktopTab.js:186]: done.

File: src/WinXP/apps/TabsMenu/index.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+  const [activeTab, setActiveTab] = useState(-1);
+
+  useEffect(() => {
+    const fallbackTabIndex = tabs.findIndex(item => isValidTab(item));
+
+    const defaultTabIndex = tabs.findIndex(
+      item => item.defaultTab && isValidTab(item),
+    );
+
+    setActiveTab(defaultTabIndex !== -1 ? defaultTabIndex : fallbackTabIndex);
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/TabsMenu/index.js:14]: Done.

File: src/WinXP/apps/TabsMenu/index.js [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    width: 100%;
+    height: 3px;
+    background-color: orange;
+    border-top-left-radius: 3px;
+    border-top-right-radius: 3px;
+    border-top: 0.5px solid grey;
+  }
+
+  &.active {
+    z-index: 100;
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/TabsMenu/index.js:95]: Done.

File: src/WinXP/apps/index.js [hunk 19]
```diff
@@ -292,6 +302,34 @@ export const appSettings = {
     maximized: window.innerWidth < 800,
     multiInstance: true,
   },
+  Properties: {
+    header: {
+      icon: empty,
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/index.js:307]: Done.

File: src/WinXP/apps/index.js [hunk 20]
```diff
@@ -292,6 +302,34 @@ export const appSettings = {
...
+    },
+    component: DisplayProperties,
+    defaultSize: {
+      width: 400,
+      height: 475,
+    },
+    defaultOffset: {
+      x: 300,
+      y: 50,
+    },
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/apps/index.js:319]: Done.

File: src/WinXP/index.js [hunk 21]
```diff
@@ -1,7 +1,8 @@
-import React, { useReducer, useRef, useCallback } from 'react';
+import React, { useReducer, useRef, useCallback, useLayoutEffect } from 'react';
 import styled, { keyframes } from 'styled-components';
 import useMouse from 'react-use/lib/useMouse';
 import ga from 'react-ga';
+import { getLocalStorage, setLocalStorage } from './utils';
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/index.js:5]: Done.

File: src/WinXP/index.js [hunk 22]
```diff
@@ -72,6 +90,7 @@ const reducer = (state, action = { type: '' }) => {
         apps,
         nextZIndex: state.nextZIndex + 1,
         focusing: FOCUSING.WINDOW,
+        contextMenuPosition: false,
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/WinXP/index.js:93]: Done.

File: src/components/ContextMenu/ContextMenuItem.js [hunk 23]
```diff
@@ -0,0 +1,134 @@
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import Properties from '../../WinXP/apps/DisplayProperties';
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/components/ContextMenu/ContextMenuItem.js:3]: Done.

File: src/components/Button/index.js [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+  height,
+  marginRight,
+  marginLeft,
+  fontSize,
+}) {
+  return (
+    <StyledButton
+      style={style}
+      onClick={onClick}
+      className={disabled ? 'disabled' : ''}
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/components/Button/index.js]: Done.

File: src/components/Button/index.js [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+      width={width}
+      marginRight={marginRight}
+      marginLeft={marginLeft}
+      fontSize={fontSize}
+      height={height}
+    >
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/components/Button/index.js]: Done.

File: src/components/Button/index.js [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+  return (
+    <StyledButton
+      style={style}
+      onClick={onClick}
+      className={disabled ? 'disabled' : ''}
+      width={width}
+      marginRight={marginRight}
+      marginLeft={marginLeft}
+      fontSize={fontSize}
+      height={height}
```
[the remark this answers is not in the document]
Comment (Tomer51m) [reply src/components/Button/index.js]: Done.

**Comment by RonnyO** [comment]
Found a bug, when choosing Windows Classic theme:
image

**Review by RonnyO — APPROVED** [review]

**Pull request merged**